File: test_merge.py
#!/usr/bin/env python3
"""Smoke tests for merge_grok_chrome_mcp.merge (no pytest required)."""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from merge_grok_chrome_mcp import CHROME_BLOCK, merge  # noqa: E402

FIXTURE = """[cli]
installer = "internal"

[mcp_servers.stitch]
command = "npx"
enabled = true

[mcp_servers.chrome-devtools]
command = "npx"
args = [
    "-y",
    "chrome-devtools-mcp@latest",
    "--autoConnect",
]
startup_timeout_sec = 120
enabled = true

[mcp_servers.github]
url = "https://example.invalid/"
enabled = true
"""


def assert_true(cond: bool, msg: str) -> None:
    if not cond:
        raise SystemExit(f"FAIL: {msg}")


def main() -> int:
    out = merge(FIXTURE)
    assert_true("[mcp_servers.stitch]" in out, "preserves other servers")
    assert_true('url = "https://example.invalid/"' in out, "preserves github")
    assert_true("chrome-devtools-mcp@latest" not in out, "replaces @latest")
    assert_true("chrome-devtools-mcp@1.8.0" in out, "pins 1.8.0")
    assert_true("--autoConnect" in out, "keeps autoConnect")
    assert_true("--screenshot-format=jpeg" in out, "jpeg flag")
    assert_true("[mcp]" in out and "max_output_bytes = 262144" in out, "raises MCP cap")
    assert_true(out.count("[mcp_servers.chrome-devtools]") == 1, "single chrome table")
    assert_true("[mcp_servers.chrome-devtools.env]" in out, "env table")
    assert_true("STITCH_API_KEY" not in out, "does not invent secrets")

    again = merge(out)
    assert_true(again == out, "idempotent")

    gone = merge(out, uninstall=True)
    assert_true("[mcp_servers.chrome-devtools]" not in gone, "uninstall strips chrome")
    assert_true("[mcp_servers.stitch]" in gone, "uninstall keeps stitch")
    assert_true("max_output_bytes = 262144" in gone, "uninstall keeps [mcp] cap")

    empty = merge("")
    assert_true(empty.startswith("[mcp]"), "empty file gets [mcp] first")
    assert_true("chrome-devtools-mcp@1.8.0" in empty, "empty file gets chrome block")
    assert_true(CHROME_BLOCK.strip() in empty, "block matches canonical")

    mcp_first = merge("[mcp]\nmax_output_bytes = 20\n\n[ui]\ntheme = \"x\"\n")
    assert_true("max_output_bytes = 262144" in mcp_first, "updates existing cap")
    assert_true('theme = "x"' in mcp_first, "keeps other [mcp]? wait this is [ui]")
    assert_true("[ui]" in mcp_first, "keeps [ui]")

    assert_true(out.endswith("\n"), "merged output ends with newline")
    assert_true(gone.endswith("\n"), "uninstall output ends with newline")

    print("ok")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: test_merge.js
#!/usr/bin/env node
"use strict";

const { merge, CHROME_BLOCK } = require("./merge_grok_chrome_mcp.js");

const FIXTURE = `[cli]
installer = "internal"

[mcp_servers.stitch]
command = "npx"
enabled = true

[mcp_servers.chrome-devtools]
command = "npx"
args = [
    "-y",
    "chrome-devtools-mcp@latest",
    "--autoConnect",
]
startup_timeout_sec = 120
enabled = true

[mcp_servers.github]
url = "https://example.invalid/"
enabled = true
`;

function assertTrue(cond, msg) {
  if (!cond) {
    console.error(`FAIL: ${msg}`);
    process.exit(1);
  }
}

const out = merge(FIXTURE);
assertTrue(out.includes("[mcp_servers.stitch]"), "preserves other servers");
assertTrue(out.includes("enabled = true"), "keeps stitch enabled");
assertTrue(out.includes('url = "https://example.invalid/"'), "preserves github");
assertTrue(!out.includes("chrome-devtools-mcp@latest"), "replaces @latest");
assertTrue(out.includes("chrome-devtools-mcp@1.8.0"), "pins 1.8.0");
assertTrue(out.includes("--autoConnect"), "keeps autoConnect");
assertTrue(out.includes("--page-id-routing"), "explicit pageId routing");
assertTrue(out.includes("--redact-network-headers"), "redacts network headers");
assertTrue(!out.includes("--allow-unrestricted-paths"), "no unrestricted paths");
assertTrue(out.includes("--screenshot-format=jpeg"), "jpeg flag");
assertTrue(out.includes("[mcp]") && out.includes("max_output_bytes = 262144"), "raises MCP cap");
assertTrue((out.match(/\[mcp_servers\.chrome-devtools\]/g) || []).length === 1, "single chrome table");
assertTrue(out.includes("[mcp_servers.chrome-devtools.env]"), "env table");
assertTrue(!out.includes("STITCH_API_KEY"), "does not invent secrets");

const again = merge(out);
assertTrue(again === out, "idempotent");

const gone = merge(out, { uninstall: true });
assertTrue(!gone.includes("[mcp_servers.chrome-devtools]"), "uninstall strips chrome");
assertTrue(gone.includes("[mcp_servers.stitch]"), "uninstall keeps stitch");
assertTrue(gone.includes("max_output_bytes = 262144"), "uninstall keeps [mcp] cap");

const empty = merge("");
assertTrue(empty.startsWith("[mcp]"), "empty file gets [mcp] first");
assertTrue(empty.includes("chrome-devtools-mcp@1.8.0"), "empty file gets chrome block");
assertTrue(empty.includes(CHROME_BLOCK.trim()), "block matches canonical");

const mcpFirst = merge('[mcp]\nmax_output_bytes = 20\n\n[ui]\ntheme = "x"\n');
assertTrue(mcpFirst.includes("max_output_bytes = 262144"), "updates existing cap");
assertTrue(mcpFirst.includes('theme = "x"'), "keeps [ui] value");
assertTrue(mcpFirst.includes("[ui]"), "keeps [ui]");
assertTrue((mcpFirst.match(/^\[mcp\]/gm) || []).length === 1, "no duplicate [mcp]");

assertTrue(out.endsWith("\n"), "merged output ends with newline");
assertTrue(gone.endsWith("\n"), "uninstall output ends with newline");

const comment = merge(`[mcp_servers.stitch]
command = "npx" # see [mcp_servers.chrome-devtools]
enabled = true
`);
assertTrue(comment.includes("enabled = true"), "comment with table name does not eat stitch");
assertTrue(comment.includes("# see [mcp_servers.chrome-devtools]"), "preserves inline comment");

const quoted = merge(`[mcp_servers.stitch]
args = [ "-y", "[mcp_servers.chrome-devtools]" ]
enabled = true
`);
assertTrue(quoted.includes('"[mcp_servers.chrome-devtools]"'), "quoted table name is not a header");
assertTrue(quoted.includes("enabled = true"), "quoted name does not eat following keys");

const crlf = merge("[mcp]\r\nmax_output_bytes = 20\r\n\r\n[ui]\r\ntheme = \"x\"\r\n");
assertTrue((crlf.match(/^\[mcp\]/gm) || []).length === 1, "CRLF does not duplicate [mcp]");
assertTrue(crlf.includes('theme = "x"'), "CRLF keeps [ui]");

const bom = merge("\uFEFF[mcp]\nmax_output_bytes = 20\n\n[ui]\ntheme = \"x\"\n");
assertTrue((bom.match(/^\[mcp\]/gm) || []).length === 1, "BOM does not duplicate [mcp]");

console.log("ok");
